pkg/operator2/status.go

```diff
@@ -1,75 +1,113 @@
 package operator2
 
 import (
-	"strings"
+	"k8s.io/apimachinery/pkg/util/sets"
 
 	operatorv1 "github.com/openshift/api/operator/v1"
 	"github.com/openshift/library-go/pkg/operator/v1helpers"
 )
 
-func handleDegraded(operatorConfig *operatorv1.Authentication, prefix string, err error) {
-	handleDegradedWithReason(operatorConfig, prefix, "Error", err)
+// Each prefix included in this set will have its condition defaulted
+// to false if not explicitly set to true.
+var degradedConditionPrefixes = sets.NewString(
+	"IdentityProviderConfig",
+	"OAuthClients",
+	"OperatorSync",
+	"RouteHealth",
+	"RouteStatus",
+	"WellKnownEndpoint",
+)
+
+type authConditions struct {
+	conditions  []operatorv1.OperatorCondition
+	hasDegraded bool
+}
+
+func newAuthConditions() *authConditions {
+	return &authConditions{
+		conditions: []operatorv1.OperatorCondition{},
+	}
 }
 
-func handleDegradedWithReason(operatorConfig *operatorv1.Authentication, prefix, reason string, err error) {
-	if err != nil {
-		v1helpers.SetOperatorCondition(&operatorConfig.Status.Conditions,
-			operatorv1.OperatorCondition{
-				Type:    prefix + operatorv1.OperatorStatusTypeDegraded,
-				Status:  operatorv1.ConditionTrue,
-				Reason:  reason,
-				Message: err.Error(),
+// all returns the set of conditions representing the complete status
+// of the auth operator. If a condition was not set, it will be
+// returned in its default state (degraded -> false, progressing ->
+// false, available -> true).
+func (c *authConditions) all() []operatorv1.OperatorCondition {
+	allConditions := []operatorv1.OperatorCondition{}
+
+	// Collect degraded conditions
+	for _, prefix := range degradedConditionPrefixes.UnsortedList() {
+		conditionType := prefix + operatorv1.OperatorStatusTypeDegraded
+		if degraded := v1helpers.FindOperatorCondition(c.conditions, conditionType); degraded != nil {
```

Comment:
**Consider**, no action required if you feel confident about the current course of action:

Only return the condition updates that were triggered throughout the course of sync loop execution and ignore their complement in the set of all controllers conditions.

Reason (by example):
checking route health fails on GET HTTP request, you switch RouteHealthDegraded on
KAS has a temporary blip so for a few next syncs you fail earlier than in the route health, turning RouteHealthDegraded to false, instead having just OperatorSyncDegraded be true. Route is still inaccessible.
KAS decides it's ok to be a good API server again, you flip OperatorSyncDegraded to false and RouteHealthDegraded back to true

Sure, a counterexample would be route goes healthy again and KAS stays degraded so we'd be reporting wrong again, but I still tend to like the first approach a bit more. It is also the same behavior we had before, whereas this would be a change.